dead_letter system table to keep erroneous incoming messages (#68873)
ClickHouse/ClickHouse: docs/en/operations/system-tables/dead_letter_queue.md

---
slug: /en/operations/system-tables/dead_letter_queue
---
# dead_letter_queue

> Contains information about messages received via a stream engine and parsed with an errors. Currently implemented for Kafka and RabbitMQ.

Reviewer: Maybe `streaming engine` is also okay.
```suggestion
Contains information about messages received via a stream like engine and parsed with an errors. Currently implemented for Kafka and RabbitMQ.
```

Author: Thanks. I think I like "streaming engine" more.

> Logging is controlled by `dead_letter_queue` of `kafka_handle_error_mode` setting.

Reviewer: ```suggestion
Logging is enabled by specifying `dead_letter_queue` for the engine specific `handle_error_mode` setting.
```

The flushing period of data is set in `flush_interval_milliseconds` parameter of the [dead_letter_queue](../../operations/server-configuration-parameters/settings.md#server_configuration_parameters-dead_letter_queue) server settings section. To force flushing, use the [SYSTEM FLUSH LOGS](../../sql-reference/statements/system.md#query_language-system-flush_logs) query.

ClickHouse does not delete data from the table automatically. See [Introduction](../../operations/system-tables/index.md#system-tables-introduction) for more details.

Columns:

> - `stream_type` ([Enum8](../../sql-reference/data-types/enum.md)) - Stream type. Possible values: 'Kafka'.

Reviewer: ```suggestion
- `stream_type` ([Enum8](../../sql-reference/data-types/enum.md)) - Stream type. Possible values: 'Kafka' or 'RabbitMQ'.
```

Author: In similar case in settings.md it is "and", not "or".
ChatGPT agrees that "and" is slightly better.

Reviewer: Yeah, that is completely fine for me as long as it is mentioned. Thanks for checking it though!

- `event_date` ([Date](../../sql-reference/data-types/date.md)) - Message consuming date.
- `event_time` ([DateTime](../../sql-reference/data-types/datetime.md)) - Message consuming date and time.
- `event_time_microseconds` ([DateTime64](../../sql-reference/data-types/datetime64.md)) - Message consuming time with microseconds precision.

> - `database_name` ([LowCardinality(String)](../../sql-reference/data-types/string.md)) - ClickHouse database Kafka table belongs to.

Reviewer: ```suggestion
- `database_name` ([LowCardinality(String)](../../sql-reference/data-types/string.md)) - ClickHouse database the table belongs to.
```
